client/setec/store.go

```diff
@@ -0,0 +1,302 @@
+	if c.PollInterval <= 0 {
+		return 1 * time.Minute
+	}
+	return c.PollInterval
+}
+
+func (c StoreConfig) cache() Cache { return c.Cache }
+
+// NewStore creates a secret store with the given configuration.  The service
+// URL of the client must be set.
+//
+// NewStore blocks until all the secrets named in cfg.Secrets are available for
+// retrieval by the Secret method, or ctx ends.  The context passed to NewStore
+// is only used for initializing the store. If a cache is provided, cached
+// values are accepted even if stale, as long as there is a value for each of
+// the secrets in cfg.
+func NewStore(ctx context.Context, cfg StoreConfig) (*Store, error) {
+	if cfg.Client.Server == "" {
+		return nil, errors.New("no service URL is set")
+	} else if len(cfg.Secrets) == 0 {
+		return nil, errors.New("no secrets are listed")
+	}
+
+	s := &Store{
+		client: cfg.Client,
+		logf:   cfg.logger(),
+		cache:  cfg.cache(),
+	}
+
+	// Initialize the active versions maps.
+	s.active.m = make(map[string]*api.SecretValue)
+	s.active.f = make(map[string]Secret)
+
+	// If we have a cache, try to load data from there first.
+	data, err := s.cache.Read()
+	if err != nil {
+		return nil, fmt.Errorf("reading cached secrets: %w", err)
+	}
+	if len(data) != 0 { // it's OK for the cache to be empty
+		if err := json.Unmarshal(data, &s.active.m); err != nil {
+			return nil, fmt.Errorf("invalid cache: %w", err)
+		}
+	}
+
+	// If there are any configured secrets that weren't cached, stub them in.
+	for _, name := range cfg.Secrets {
+		if _, ok := s.active.m[name]; !ok {
+			s.active.m[name] = nil
+		}
+	}
+
+	// Ensure we have values for all requested secrets.
+	if err := s.initializeActive(ctx); err != nil {
+		return nil, err
+	}
+
+	// Start a background task to refresh secrets.
+	pctx, cancel := context.WithCancel(context.Background())
+	s.cancel = cancel
+
+	done := make(chan struct{})
+	s.done = done
+
+	go s.run(pctx, cfg.pollInterval(), done)
+
+	return s, nil
+}
+
+// Close stops the background task polling for updates and waits for it to
+// exit.
+func (s *Store) Close() error {
+	s.cancel()
+	<-s.done
+	return nil
+}
+
+// Secret returns a fetcher for the named secret. It returns nil if name does
+// not correspond to one of the secrets known by s.
+func (s *Store) Secret(name string) Secret {
+	s.active.Lock()
+	defer s.active.Unlock()
+	if _, ok := s.active.m[name]; !ok {
+		return nil // unknown secret
+	}
+	f, ok := s.active.f[name]
+	if !ok {
+		f = func() []byte {
+			s.active.RLock()
+			defer s.active.RUnlock()
+			return s.active.m[name].Value
+		}
+		s.active.f[name] = f
+	}
+	return f
+}
+
+// A Secret is a function that fetches the current active value of a secret.
+// The caller should not cache the value returned; the function does not block
+// and will always report a valid (if possibly stale) result.
+type Secret func() []byte
+
+// Get returns the current active value of the secret.  It is a shorthand for
```

Comment:
More like a longhand for calling the function :zany_face:

Reply:
Haha, fair enough. Updated.